text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,182 @@
+    uint8_t a;
+    uint32_t b;
+};
+```
+
+Currently, `#[repr(packed(_))]` structs cannot be `#[repr(align(_))]` or transitively contain `#[repr(align(_))]` types. Attempting to do so results in a [hard error](https://doc.rust-lang.org/nightly/error_codes/E0588.html).
+
+This behavior was added in the [original implementation](https://github.com/rust-lang/rust/issues/33158) of `#[repr(packed)]` due to concerns over differing behavior between msvc and gcc/clang. This makes it cumbersome or even impossible to produce C-compatible struct layouts in Rust when the corresponding C types were annotated with both `packed` and `aligned`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## `#[repr(C)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target toolchain ABI.
+
+For example, given:
+```c
+#[repr(C, align(4))]
+struct Foo(u8);
+#[repr(C, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.
+
+
+## `#[repr(system)]`
+When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
+the resulting layout matches the target OS ABI.
+
+For example, given:
+```c
+#[repr(system, align(4))]
+struct Foo(u8);
+#[repr(system, packed(1))]
+struct Bar(Foo);
+```
+`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and `*-pc-windows-gnu`. It would be 1 for everything else.
+
+## `#[repr(Rust)]`
+When `align(N)` and `packed(M)` attributes exist on the same type, or when `packed` structs contain `aligned` fields,
+the type will have their base alignment increased to `N`, while the struct fields will be laid out as if their
+alignments were decreased to `M`. However, in general Rust is free to reorder
+these fields for optimization purposes, and the only guarantee is that the fields will maintain a minimum alignment of `M`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In the following paragraphs, "Decreasing M to N" means:
+```
+if M > N {
+    M = n
+}
+```
+
+"Increasing M to N" means:
+```
+if M < N {
+    M = N
+}
+```
+
+
+`#[repr(align(N))]` increases the base alignment of a type to be N.
+
+`#[repr(packed(M))]` decreases the alignment of the struct fields to be M. Because the base alignment of the type
+is defined as the maximum of the alignment for any fields, this also has the indirect result of decreasing the base
+alignment of the type to be M.
+
+When the align and packed modifiers are applied on the same type as `#[repr(align(N), packed(M))]`,
+the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
+increased to be N.
+
+When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type, depending on the
+target triplet, either:
```

Comment:
This should IMO be rephrased as an algorithm that works one "layer" at a time. Computing the layout of a type T should only consider the fields of T and their properties. It should never recurse into the fields of T.

I don't currently actually understand the proposed spec here, and this should make it a lot easier to understand.

Reply:
My current understanding of the algorithm is as follows:

---

We equip types with a new property, the "explicitly requested alignment". For all base types, this alignment is 1. For structs, it is by default the maximum of the explicitly requested alignments of all fields. For a struct with the `#[repr(align(N))]` attribute, the explicitly requested alignment the maximum of N and the explicitly requested alignments of its fields.

Note that the explicitly requested alignment of a type can never be bigger than the required alignment of the type.

When computing the layout of a packed(P) struct, then currently we ensure each field is aligned to `min(A, P)`, where `A` is the (regular) alignment of the field type. Under the new rules, *only on MSVC targets*, we instead ensure the field is aligned to `max(E, min(A, P))`, where `E` is the explicitly requested alignment. (Due to the aforementioned inequality, this is equivalent to `min(max(E, P), A)`.) This is the only time the explicitly requested alignment of a type has any effect.

---

I am not fully confident that this is correct. Here's a corner case:
```rust
#[repr(C, align(4))]
struct Align4(i32);

#[repr(C, align(2))]
struct Align2(Align4);

#[repr(C, packed)]
struct Packed(Align2);
```
What is the resulting alignment of `Packed` on MSVC? My proposed algorithm says 4.  Is that correct?

Reply:
Note: `alignas` on the `struct` is allowed in standard C++ but not standard C. It's a common extension to allow using `alignas` in this position in C anyway. [[cppeference]](https://en.cppreference.com/w/c/language/_Alignas)

```c
#include <stdalign.h>
#include <stddef.h>
#incluse <stdint.h>
#include <stdio.h>

struct alignas(4) Align4
{
    int32_t _0;
};

struct alignas(2) Align2
{
    Align4 _0;
};

#pragma pack(push, 1)
struct Packed
{
    Align2 _0;
};
#pragma pack(pop)

struct Group
{
    uint8_t _0;
    Align4 _1;
};

#pragma pack(push, 1)
struct P
{
    uint16_t _0;
    Group _1;
};
#pragma pack(pop)

int main()
{
    printf("alignof(Packed) = %zu\n", alignof(Packed));
    printf("\n");
    printf("offsetof(P, _0) = %zu\n", offsetof(P, _0));
    printf("offsetof(P, _1) = %zu\n", offsetof(P, _1));
    printf("offsetof(P, _1._0) = %zu\n", offsetof(P, _1) + offsetof(Group, _0));
    printf("offsetof(P, _1._1) = %zu\n", offsetof(P, _1) + offsetof(Group, _1));
}
```

(I wrote this from mobile, so I'm moving to my desktop to actually test this now. IIUC, `Align2` should be an error on MSVC, and you have the layouts otherwise correct. Time to test…)

Reply:
Oh, so having an `align` attribute on a struct where a field already has a higher explicitly requested alignment is an error? Should this also be an error in Rust? The RFC doesn't say so, and it would be a breaking change... but that should at least be mentioned in the RFC. It might be worth a warning if someone writes a `repr(C)` type in Rust that couldn't be written in C.

Reply:
https://c.godbolt.org/z/es1cGPhz9

On MSVC 19.40 (VS 17.10) in C mode,

```c
#include <stdalign.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>

__declspec(align(4))
struct Align4
{
    int32_t _0;
};

__declspec(align(2))
struct Align2
{
    struct Align4 _0;
};

#pragma pack(push, 1)
struct Packed
{
    struct Align2 _0;
};
#pragma pack(pop)

struct Group
{
    uint8_t _0;
    struct Align4 _1;
};

#pragma pack(push, 1)
struct P
{
    uint16_t _0;
    struct Group _1;
};
#pragma pack(pop)

int main()
{
    printf("alignof(Packed) = %zu\n", alignof(struct Packed));
    printf("\n");
    printf("offsetof(P, _0) = %zu\n", offsetof(struct P, _0));
    printf("offsetof(P, _1) = %zu\n", offsetof(struct P, _1));
    printf("offsetof(P, _1._0) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _0));
    printf("offsetof(P, _1._1) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _1));
}
```

gives

```text
alignof(Packed) = 4

offsetof(P, _0) = 0
offsetof(P, _1) = 4
offsetof(P, _1._0) = 4
offsetof(P, _1._1) = 8
```

Using `alignas`/`_Alignas` on the type (`alignas(N) struct Tag`) instead of `__declspec(align)` gives

> warning C5274: behavior change: _Alignas no longer applies to the type 'Align4' (only applies to declared data objects)

and results of 1 / 0, 2, 2, 6; fully just ignoring the `alignas` modifier. Note that standard C does not permit the use of `alignas` for struct definitions. C++ does. In C++ mode (`/std:c++latest`, using `struct alignas(N) Tag`), MSVC gives:

```text
alignof(Packed) = 4

offsetof(P, _0) = 0
offsetof(P, _1) = 4
offsetof(P, _1._0) = 4
offsetof(P, _1._1) = 8
```

along with a warning:

> warning C4359: 'Align2': Alignment specifier is less than actual alignment (4), and will be ignored.

Reply:
For completeness, clang *does not* seem to implement this fully, unless I made a mistake: [[godbolt]](https://c.godbolt.org/z/rTGdazbxE)

<details>

```c
#include <stdalign.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>

struct [[gnu::ms_struct]] Align4
{
    alignas(4)
    int32_t _0;
};

struct [[gnu::ms_struct]] Align2
{
    // alignas(2) // error: requested alignment is less than minimum alignment of 4 for type 'struct Align4'
    struct Align4 _0;
};

struct [[gnu::packed]] Packed
{
    struct Align2 _0;
};

struct Group
{
    uint8_t _0;
    struct Align4 _1;
};

struct [[gnu::packed]] P
{
    uint16_t _0;
    struct Group _1;
};

int main()
{
    printf("alignof(Packed) = %zu\n", alignof(struct Packed));
    printf("\n");
    printf("offsetof(P, _0) = %zu\n", offsetof(struct P, _0));
    printf("offsetof(P, _1) = %zu\n", offsetof(struct P, _1));
    printf("offsetof(P, _1._0) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _0));
    printf("offsetof(P, _1._1) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _1));
}
```

```text
alignof(Packed) = 1

offsetof(P, _0) = 0
offsetof(P, _1) = 2
offsetof(P, _1._0) = 2
offsetof(P, _1._1) = 6
```

</details>